runtime/src/rad_claims.rs

```diff
@@ -0,0 +1,494 @@
+        /// Map of root hashes that correspond to lists of RAD reward claim amounts per account.
+        RootHashes get(fn get_root_hash): map hasher(blake2_128_concat) T::Hash => bool;
+        /// Account that is allowed to upload new root hashes.
+        UploadAccount get(fn get_upload_account): T::AccountId;
+    }
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait>{
+        /// The combination of account id, amount, and proofs vector in a claim was invalid.
+        InvalidProofs,
+        /// The payout amount attempting to be claimed is less than the minimum allowed by [MIN_PAYOUT].
+        UnderMinPayout,
+        /// Amount being claimed is less than the available amount in [AccountBalances].
+        InsufficientBalance,
+        /// Protected operation, must be performed by admin
+        MustBeAdmin,
+    }
+}
+
+decl_event! {
+    pub enum Event<T> where
+        <T as frame_system::Trait>::AccountId,
+        <T as frame_system::Trait>::Hash,
+        <T as pallet_balances::Trait>::Balance,
+    {
+        Claimed(AccountId, Balance),
+        RootHashStored(Hash),
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        fn deposit_event() = default;
+
+        /// Claims RAD tokens awarded through tinlake investments
+        /// Feeless dispatchable function
+        /// The extrinsic is validated by the custom `validate_unsigned` function below
+        ///
+        /// # <weight>
+        /// - Based on hashes length
+        /// # </weight>
+        #[weight = 195_000_000]
+        pub fn claim(origin,
+                     account_id: T::AccountId,
+                     amount: T::Balance,
+                     sorted_hashes: Vec<T::Hash>,
+        ) -> DispatchResult {
+            ensure_none(origin)?;
+
+            let claimed = Self::get_account_balance(&account_id);
+
+            // Payout = amount - claim
+            let payout = amount.checked_sub(&claimed)
+                .ok_or(Error::<T>::InsufficientBalance)?;
+
+            // Payout must not be less than minimum allowed
+            ensure!(payout >= MIN_PAYOUT.saturated_into(),
+                    Error::<T>::UnderMinPayout);
+
+            let source = MODULE_ID.into_account();
+            // Checking balance of Module account before proceeding
+            let current_balance = T::Currency::free_balance(&source);
+            ensure!(U256::from(current_balance.saturated_into()) >= U256::from(payout.saturated_into()), Error::<T>::InsufficientBalance);
+
+            // Set account balance to amount
+            AccountBalances::<T>::insert(account_id.clone(), amount);
+
+            // Transfer payout amount
+            <pallet_balances::Module<T> as Currency<_>>::transfer(
+                &source,
+                &account_id,
+                payout,
+                AllowDeath,
```

Comment:
Not sure we want to `AllowDeath` for the source. Some dust amount might be lost due to this

Reply:
Yep since it is the keyless account, I think it makes sense to KeepAlive. I would have to add a check above to include the existential_deposit amount in the balance check above.